File: go/pricing/pricingcontext_test.go
// ABOUTME: Unit tests for the PricingContext type.
// ABOUTME: Verifies context creation and tag checking.
package pricing

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestNewPricingContext(t *testing.T) {
	now := time.Date(2025, 12, 18, 10, 0, 0, 0, time.UTC)
	channel := "online"
	customerID := "CUST123"
	customerTags := map[string]bool{"new": true, "gold": true}

	context := NewPricingContext(now, channel, customerID, customerTags)

	assert.Equal(t, now, context.Now)
	assert.Equal(t, channel, context.Channel)
	assert.Equal(t, customerID, context.CustomerID)
	assert.Equal(t, customerTags, context.CustomerTags)
}

func TestPricingContextHasTag(t *testing.T) {
	now := time.Date(2025, 12, 18, 10, 0, 0, 0, time.UTC)
	context := PricingContext{
		Now:          now,
		Channel:      "online",
		CustomerID:   "CUST123",
		CustomerTags: map[string]bool{"new": true, "gold": true},
	}

	assert.True(t, context.HasTag("new"))
	assert.True(t, context.HasTag("gold"))
	assert.False(t, context.HasTag("vip"))
}

File: typescript/src/pricing/PricingContext.test.ts
import { describe, it, expect } from 'vitest';
import { PricingContext } from './PricingContext';

describe('PricingContext', () => {
  describe('construction', () => {
    it('should construct with valid fields', () => {
      const now = new Date('2025-12-18T10:00:00Z');
      const tags = new Set(['new', 'gold']);
      const context = new PricingContext(now, 'online', 'CUST123', tags);

      expect(context.now).toBe(now);
      expect(context.channel).toBe('online');
      expect(context.customerId).toBe('CUST123');
      expect(context.customerTags).toBe(tags);
    });
  });

  describe('hasTag', () => {
    it('should return true for existing tag', () => {
      const tags = new Set(['new', 'gold']);
      const context = new PricingContext(new Date(), 'online', 'CUST123', tags);

      expect(context.hasTag('new')).toBe(true);
      expect(context.hasTag('gold')).toBe(true);
    });

    it('should return false for non-existing tag', () => {
      const tags = new Set(['new', 'gold']);
      const context = new PricingContext(new Date(), 'online', 'CUST123', tags);

      expect(context.hasTag('silver')).toBe(false);
      expect(context.hasTag('platinum')).toBe(false);
    });

    it('should work with empty tags', () => {
      const tags = new Set<string>();
      const context = new PricingContext(new Date(), 'online', 'CUST123', tags);

      expect(context.hasTag('new')).toBe(false);
    });
  });
});
